pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -65,23 +67,22 @@ public class PersistentStickyKeyDispatcherMultipleConsumers extends PersistentDi
     private final KeySharedMode keySharedMode;
 
     /**
-     * When a consumer joins, it will be added to this map with the current read position.
-     * This means that, in order to preserve ordering, new consumers can only receive old
-     * messages, until the mark-delete position will move past this point.
+     * When a consumer joins, it will be added to this map with the current last sent position per the message key.
+     * This means that, in order to preserve ordering per the message key, new consumers can only receive old
+     * messages, until the mark-delete position will move past this point in the key. New consumers can receive
+     * any new messages with the message key that is not in the last sent position.
      */
-    private final LinkedHashMap<Consumer, PositionImpl> recentlyJoinedConsumers;
-
-    private final Set<Consumer> stuckConsumers;
-    private final Set<Consumer> nextStuckConsumers;
+    private final LinkedHashMap<Consumer, LastSentPositions> recentlyJoinedConsumers;
+    // The lastSentPosition is not thread-safe
+    private final LastSentPositions lastSentPositions;
 
     PersistentStickyKeyDispatcherMultipleConsumers(PersistentTopic topic, ManagedCursor cursor,
```

Comment:
Hi @equanz 

Sorry, I've been so busy lately.

### Menu
- Explain the issue.
- Suggestion a simpler solution instead of the current one

--- 

### Explain the issue

There have two scenarios that will make consumption out of order, which this PR tries to fix.

#### Scenario-1

| time | `process: add consumer` | `process: delivery messages to client` |
| --- | --- | --- |
| 1 |  | Read entries `1:6 ~ 1:10` complete |
| 2 | Add new consumer into the selector |
| 3 | | Choose consumer by the selector |
| 4 | Add the new consumer into recently joined consumers | 
| 5 |  | Delivery entries `1:6 ~ 1:10` to all consumers(includes old consumers and the new consumer) |

#### Scenario-2

| time | `process: add consumer` | `process: delivery messages to client` |
| --- | --- | --- |
| 1 | read position is `1:6` |
| 2 | Add new consumer into the selector |
| 3 |  | Read entries `1:6 ~ 1:10` complete |
| 4 |  | Set read position `1:11` |
| 5 | Add the new consumer into recently joined consumers | 
| 6 | <strong>(Highlight)</strong>The max read-position of the new consumer is `1:11`, but the exact correct value is `1:6` | 
| 7 | | Choose consumer by the selector |
| 8 |  | Delivery entries `1:6 ~ 1:10` to all consumers(includes old consumers and the new consumer) |

--- 

### A simpler solution

First of all, this is a great catch, and the current patch can solve the issues above.

But this patch makes an already complicated mechanism even more complicated, and I'd like to suggest a simple solution:  Change the constructor of `recentlyJoinedConsumers` richer, let me show a simple code:

```java
 /**
 * There have two positions in the value:
 * The first position:  the read position of the cursor when the consumer joins.
 * The second position: the first entry in normal reading after the consumer joined.
 */
protected Map<Consumer, Pair<PositionImpl,PositionImpl> recentlyJoinedConsumers;

/**
 * Get max read the position of the consumer.
 */
protected PositionImpl getMaxReadPosition(Consumer consumer) {
    Pair<PositionImpl,PositionImpl> pair = recentlyJoinedConsumers.get(consumer);
    return Math.min(pair.getKey(), pair.getValue());
}

/**
 * Register consumer.
 */
public synchronized CompletableFuture<Void> addConsumer(Consumer consumer) {
    addConsumerIntoRecentJoinQueue();
    setReadPositionWhenConsumerJoined();
    // do other things.
}

/**
 * Calculate how many entries the consumer would consume.
 */
private int getRestrictedMaxEntriesForConsumer(Consumer consumer, List<Entry> entries, ReadType readType){
   setFirstReadPositionOfRecentJoinedConsumers(entries.get(0));

   PositionImpl maxReadPosition = getMaxReadPosition(consumer);
    ...
}
```

What do you think of this solution?

Reply:
Thank you for your deep insight!
I have a question about your proposed solution.

> The second position: the first entry in normal reading after the consumer joined.

Could you explain this position in more detail, please? When and how to get the position?
(Perhaps, this solution is similar to https://github.com/apache/pulsar/pull/20179#issuecomment-1539314184?)

Reply:
@equanz 

> Could you explain this position in more detail, please? When and how to get (or calculate?) the position?

It will be set in method `getRestrictedMaxEntriesForConsumer`, let me show a simpler code

```java
/**
 * Calculate how many entries the consumer would consume.
 * @param theFirstEntryInCurrentReading This is not the first message the consumer will receive, but the first message read in this batch.
 */
private synchronized int getRestrictedMaxEntriesForConsumer(Consumer consumer, List<Entry> entries, ReadType readType, PositionImpl theFirstEntryInCurrentReading){

   if (readType.equals(Normal) ){
       recentlyJoinedConsumers.get(consumer).getRight() == null;
       recentlyJoinedConsumers.get(consumer).getRight() = theFirstEntryInCurrentReading;
   }

   PositionImpl maxReadPosition = getMaxReadPosition(consumer);
    ...
}
```

The progress runs like this:

| time | `process: add consumer` | `process: in-flight reading` |
| --- | --- | --- |
| 1 | read position is `1:6` |
| 2 | | an in-flight reading which tries to read `1:6~1:10` |
| 3 | Add the new consumer into recently joined consumers | 
| 4 | Add new consumer into the selector |
| 5 |  | Read entries `1:6 ~ 1:10` complete |
| 6 |  | Set read position `1:11` |
| 7 | Set `recentlyJoinedConsumers.consumer.first-position` to `1:11` |
| 8 | | Try to send messages to consumers |
| 9 | | Set `recentlyJoinedConsumers.consumer.second-position` to `1:6`|

Reply:
@equanz Sorry, my reply has been delayed again

Let me summarize this scenario:
1. a message(`Msg_1`) was put into the Replay Queue caused by the target consumer(`C_1`) having no permits.
2. at this moment, the new Consumer(`C_2`) takes over the key of `Msg_1` above.
3. in the next round, `Msg_1` was delivered to `C_2`.

Is this explain correct?<sup>[1]</sup>

If yes<sup>[1]</sup>, I have some question:
- If we revert the PR https://github.com/apache/pulsar/pull/20026, the `step 1` and `step 2`  can not be executed concurrently, right?<sup>[2]</sup>

If yes <sup>[2]</sup>, there have 2 case:
#### Case 1: `step 1` runs before `step 2`
In this case, the `max read position` of `C_2` should be larger than `Msg-1`, then `Msg_1` will not be sent to `C_2` in the next round.

#### Case 1: `step 2` runs before `step 1`
In this case, the key of `Msg_1` above will still be taken over by `C_1`, then `Msg_1` will not be sent to `C_2` in the next round.

Reply:
Hi @poorbarcode

* [1] I think it is not entirely correct. Your scenario step seems roughly right, but based on my analysis, the root cause of the issue is not related to concurrency.
  The cause lies in [the isDispatcherStuckOnReplays feature](https://github.com/apache/pulsar/blob/b7f0004313ea4565717cc6d3c0b99aee5c079c6c/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java#L429-L434), and [the recently joined consumers feature](https://github.com/apache/pulsar/blob/b7f0004313ea4565717cc6d3c0b99aee5c079c6c/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java#L371-L377).

Reply:
@equanz I think this is a good way to solve the above problems and I have several suggestions
- Since the current mechanism of Key_Shared mode is complex enough. hopefully, a PIP can be proposed for this change
- Add a mechanism to limit the memory cost of the mapping of the key and last sent position.

Reply:
Sorry for too late.
I opened a new PIP design PR.

https://github.com/apache/pulsar/pull/20776

Reply:
@equanz Thanks